src/sources/kafka.rs

```diff
@@ -346,14 +357,34 @@ fn create_consumer(
     }
 
     let consumer = client_config
-        .create_with_context::<_, StreamConsumer<_>>(KafkaStatisticsContext)
+        .create_with_context::<_, StreamConsumer<_>>(CustomContext::default())
         .context(KafkaCreateError)?;
     let topics: Vec<&str> = config.topics.iter().map(|s| s.as_str()).collect();
     consumer.subscribe(&topics).context(KafkaSubscribeError)?;
 
     Ok(consumer)
 }
 
+#[derive(Default)]
+struct CustomContext {
+    stats: kafka::KafkaStatisticsContext,
+    finalizer: OnceCell<Arc<OrderedFinalizer<FinalizerEntry>>>,
+}
+
+impl ClientContext for CustomContext {
+    fn stats(&self, statistics: Statistics) {
+        self.stats.stats(statistics)
+    }
+}
+
+impl ConsumerContext for CustomContext {
+    fn post_rebalance(&self, _rebalance: &Rebalance) {
+        if let Some(finalizer) = self.finalizer.get() {
+            finalizer.flush();
+        }
+    }
+}
```

Comment:
👍 this _should_ work but I'm wondering if we could make it slightly more efficient by only dropping the finalizers for partitions we aren't the consumer of anymore? If that's complicated, I'm ok with this change as-is; it just may result in additional duplication in sinks.

Reply:
I thought about that, yes. I think it's doable, but could get complicated due to which piece has what data. I also had some doubts if doing that would still leave a hole due to the scenario outlined in https://github.com/vectordotdev/vector/issues/9587#issuecomment-946850281, where we end up getting the same partition again but the offsets aren't valid any more. Of course, then we should receive more than one rebalance event, so we should have dropped the finalizers anyways.

This may also only be relevant in a peculiar set of circumstances, since the next acknowledged batch of data ends up marking everything before it as done. As such, it would only matter during the time until the next piece of data for the partition is sent and acknowledged.

Reply:
Yeah, I can see how it would be complicated to track. This is a definite step forward so I'm fine with it as-is.